presto-main/src/main/java/com/facebook/presto/operator/Driver.java

```diff
@@ -324,10 +325,20 @@ private void processNewSource(TaskSource source)
                 // this state change by calling isFinished
                 return NOT_BLOCKED;
             }
+            revokeMemoryIfNecessary();
             return processInternal();
         }
     }
 
+    private void revokeMemoryIfNecessary()
+    {
+        if (driverContext.isWaitingForRevocableMemory()) {
+            for (Operator operator : operators) {
+                operator.revokeMemory();
```

Comment:
I think you should check after each revoke that we are still waiting for revokable memory